### PR TITLE
1.11 Update (2)

### PR DESCRIPTION
**Pull request by Belsnickle**
Second attempt.  Used change log to fix all instances of .isNull correctly.  One thing I am not sure about is how I handle subMeshList in BatchedGeometry.cpp.  Example1 now exits without a crash.

After I closed my original pull request for my fork the OGRECave repository was updated for MINGW builds.  I tried committing the changes to my fork but I ended up commits ahead and behind and not "even".  I ended up deleting my fork and then forking again.  Is there something I am missing to update forks to keep the "even" tag?

### CODE REVIEW

**Comment by Belsnickle** [comment]
Sorry the try and catch statement should not be in BatchedGeometry::clear().  The function should remain  unchanged from the master branch

**Comment by Belsnickle** [comment]
While moving through the forest I encountered a crash.  I have been able to reproduce this crash by moving around the one of the four edges of the forest created in Example1.  Interestingly this crash doesn't seem to happen when supplying 0 for the y position of the trees instead of using the getTerrainHeight fuction.  Ogre::Node->setPosition asserts "Invalid vector supplied as parameter".  Traced back to line 347 of BatchedGeometry.cpp.

`m_pSceneNode = m_pParentSceneNode->createChildSceneNode(m_vecCenter);`

When watching the variables the value is {...} in Visual Studio.  What does this mean?

Call stack:
```
>	Project2.exe!Forests::BatchedGeometry::build() Line 347	C++
 	Project2.exe!Forests::BatchPage::build() Line 127	C++
 	Project2.exe!Forests::GeometryPageManager::_loadPage(Forests::GeometryPage * page) Line 879	C++
 	Project2.exe!Forests::GeometryPageManager::update(unsigned long deltaTime, Ogre::Vector<3,float> & camPos, Ogre::Vector<3,float> & camSpeed, bool & enableCache, Forests::GeometryPageManager * prevManager) Line 574	C++
 	Project2.exe!Forests::PagedGeometry::update() Line 247	C++
 	Project2.exe!rayTech::frameStarted(const Ogre::FrameEvent & evt) Line 240	C++
 	OgreMain_d.dll!Ogre::Root::_fireFrameStarted(Ogre::FrameEvent & evt) Line 723	C++
 	OgreMain_d.dll!Ogre::Root::_fireFrameStarted() Line 785	C++
 	OgreMain_d.dll!Ogre::Root::renderOneFrame() Line 873	C++
 	Project2.exe!WinMain(HINSTANCE__ * hInst, HINSTANCE__ * __formal, char * strCmdLine, int __formal) Line 22	C++
```

**Comment by paroj** [comment]
> After I closed my original pull request for my fork the OGRECave repository was updated for MINGW builds. I tried committing the changes to my fork but I ended up commits ahead and behind and not "even". I ended up deleting my fork and then forking again. Is there something I am missing to update forks to keep the "even" tag?

```
git fetch origin
git reset --hard origin/master
git push -f
```
however this will discard all your local changes. Read up `git rebase` for keeping your changes.

File: source/BatchedGeometry.cpp [hunk 1]
```diff
@@ -212,16 +212,17 @@ uint32 CountUsedVertices(IndexData *id, std::map<uint32, uint32> &ibmap)
...
-   Mesh::SubMeshIterator subMeshIterator = mesh->getSubMeshIterator();
+   Mesh::SubMeshList subMeshList = mesh->getSubMeshes();
 
    // Get shared vertex data
    VertexData *oldVertexData = mesh->sharedVertexData;
 
-   while (subMeshIterator.hasMoreElements()) {
-      SubMesh *subMesh = subMeshIterator.getNext();
+   for (auto subMeshIter = subMeshList.begin(); subMeshIter != subMeshList.end(); ++subMeshIter) {
+	   SubMesh *subMesh = subMeshList.back();
```
Comment (paroj) [inline source/BatchedGeometry.cpp]: this is better as you are now actually iterating over the list, but still wrong as you clear it at that.

use: `for(auto subMesh : submeshList)`

File: source/BatchedGeometry.cpp [hunk 2]
```diff
@@ -365,16 +366,22 @@ void BatchedGeometry::build()
...
-
-      m_pSceneNode = 0;
-   }
+	if (m_pSceneNode)
+	{
+		try
+		{
+			m_pSceneNode->removeAllChildren();
+		}
+		catch (const std::exception& e)
```
Comment (paroj) [inline source/BatchedGeometry.cpp]: this is an unrelated change and should be explained - especially if you silently swallow the exception

File: source/StaticBillboardSet.cpp [hunk 3]
```diff
@@ -705,8 +705,11 @@ MaterialPtr StaticBillboardSet::getFadeMaterial(const Ogre::MaterialPtr &protoMa
             GpuProgramParametersSharedPtr params = pass->getVertexProgramParameters();
 
             //glsl can use the built in gl_ModelViewProjectionMatrix
-            if (!isglsl)
-               params->setNamedAutoConstant("worldViewProj", GpuProgramParameters::ACT_WORLDVIEWPROJ_MATRIX);
+			if (!isglsl)
+			{
+				static const Ogre::String worldViewProj = "worldViewProj";
```
Comment (paroj) [inline source/StaticBillboardSet.cpp]: seems like an unrelated/ unnecessary change

File: source/ImpostorPage.cpp [hunk 4]
```diff
@@ -445,22 +445,22 @@ void ImpostorTexture::updateMaterials()
...
 	//Delete textures
-	assert(!texture.isNull());
+	assert(texture);
 	String texName(texture->getName());
 		
-	texture.setNull();
+	texture.reset();
 	if (TextureManager::getSingletonPtr())
-		TextureManager::getSingleton().remove(texName);
+		TextureManager::getSingleton().remove(texName, RGN_AUTODETECT);
```
Comment (paroj) [inline source/ImpostorPage.cpp]: we could actually just remove it by ptr instead of by name. But your change certainly improves things

**Comment by paroj** [comment]
> Interestingly this crash doesn't seem to happen when supplying 0 for the y position of the trees instead of using the getTerrainHeight fuction.

while this is unrelated to this PR, the assertion is triggered as you supply NaN for the y postion. You probably commented out the (broken) terrain loading in Sample1. Therefore there is no intersection.

File: source/StaticBillboardSet.cpp (hunk 3 above)
Comment (Belsnickle) [reply source/StaticBillboardSet.cpp]: I was getting an error running the samples that worldViewProj wasn't a valid parameter.  I read that this is due to compiler optimization removing the "unused" worldViewProj.  This change allowed the plugin to run.

File: source/BatchedGeometry.cpp (hunk 2 above)
Comment (Belsnickle) [reply source/BatchedGeometry.cpp]: Remnant of debugging the clear() method.  I have since removed the try/catch statement.

**Comment by Belsnickle** [comment]
> while this is unrelated to this PR, the assertion is triggered as you supply NaN for the y postion. You probably commented out the (broken) terrain loading in Sample1. Therefore there is no terrain intersection.

I changed the first sample to use a volume terrain.  

When running position.y = 0 (for tree height) I can't reproduce a crash.  When using  position.y = getTerrainHeight(x,z) the trees are put on the terrain and when going around the edges of the forest it crashes every time in a certain spot near one of the four corners.

**Comment by paroj** [comment]
manually merged the uncritical parts.
